lib/Dialect/LLHD/Transforms/Mem2Reg.cpp

```diff
@@ -1255,12 +1313,96 @@ void Promoter::resolveDefinitions(ProbeNode *node) {
     return;
   auto *def = node->valueBefore->reachingDefs.lookup(blockingSlot(node->slot));
   assert(def && "no definition reaches probe");
+  auto value = def->getValueOrPlaceholder();
+
+  // Replace the probe result with the value.
   LLVM_DEBUG(llvm::dbgs() << "- Replacing " << *node->op << "\n");
-  replaceValueWith(node->op->getResult(0), def->getValueOrPlaceholder());
+  replaceValueWith(node->op->getResult(0), value);
+  eraseLaterIfUnused(node->op->getOperands());
   node->op->erase();
   node->op = nullptr;
 }
 
+/// Mutate reaching definitions for conditional drives.
+void Promoter::resolveDefinitions(DriveNode *node) {
+  if (!slotOrder.contains(getSlot(node->slot)))
+    return;
+
+  // Check whether the drive already has a value and condition set for its
+  // reaching def. Conditional drives will have no value set initially, in which
+  // case either the value is null or it is a placeholder.
+  if (!node->def->value || node->def->valueIsPlaceholder)
+    resolveDefinitionValue(node);
+  if (node->def->condition.isConditional() &&
+      (!node->def->condition.getCondition() ||
+       node->def->conditionIsPlaceholder))
+    resolveDefinitionCondition(node);
+}
+
+void Promoter::resolveDefinitionValue(DriveNode *node) {
+  // Get the slot definition reaching the drive. This is the value the drive has
+  // to update.
+  auto slot = getSlot(node->slot);
+  auto *inDef = node->valueBefore->reachingDefs.lookup(blockingSlot(slot));
+  assert(inDef && "no definition reaches drive");
```

Comment:
This is guarded by the `if (!slotOrder.contains(getSlot(node->slot))) return;` above to not be reached by non-blocking drives, right?

Reply:
Oh excellent catch. This should actually be `node->slot`, not `blockingSlot(slot)`. I'll add a test for that.

I think the `!slotOrder.contains(...)` can actually be dropped, because the pass only creates lattice nodes for drives that target promotable slots. So this check will always pass.